sway-core/src/type_system/id.rs

```diff
@@ -317,16 +325,108 @@ impl TypeId {
-            found.insert(*self);
+            found.insert(*self, trait_constraints);
         }
         found
     }
+
+    /// `check_type_parameter_bounds` does two types of checks. Lets use the example below for demonstrating the two checks:
+    /// ```ignore
+    /// enum MyEnum<T> where T: MyAdd {
+    ///   X: T,
+    /// }
+    /// ```
+    /// The enum above has a constraint where `T` should implement the trait `MyAdd`.
+    ///
+    /// If `check_type_parameter_bounds` is called on type `MyEnum<u64>` and `u64`
+    /// does not implement the trait `MyAdd` then the error `CompileError::TraitConstraintNotSatisfied`
+    /// is thrown.
+    ///
+    /// The second type of check performed results in an error for the example below.
+    /// ```ignore
+    /// fn add2<G>(e: MyEnum<G>) -> G {
+    /// }
+    /// ```
+    /// If `check_type_parameter_bounds` is called on type `MyEnum<G>` and the type parameter `G`
+    /// does not have the trait constraint `where G: MyAdd` then the error `CompileError::TraitConstraintMissing`
+    /// is thrown.
+    pub(crate) fn check_type_parameter_bounds(
+        &self,
+        ctx: &TypeCheckContext,
+        span: &Span,
+    ) -> CompileResult<()> {
+        let warnings = vec![];
+        let mut errors = vec![];
+        let engines = ctx.engines();
+
+        let structure_generics = engines
+            .te()
+            .get(*self)
+            .extract_inner_types_with_trait_constraints(engines);
+
+        for (structure_type_id, structure_trait_constraints) in &structure_generics {
+            if structure_trait_constraints.is_empty() {
+                continue;
+            }
+
+            let structure_type_info = engines.te().get(*structure_type_id);
+            let structure_type_info_with_engines = engines.help_out(structure_type_info.clone());
+            if let TypeInfo::UnknownGeneric {
+                trait_constraints, ..
+            } = &structure_type_info
+            {
+                let mut generic_trait_constraints_trait_names: Vec<CallPath<BaseIdent>> = vec![];
+                for trait_constraint in trait_constraints.iter() {
+                    generic_trait_constraints_trait_names.push(trait_constraint.trait_name.clone());
+                }
+                for structure_trait_constraint in structure_trait_constraints {
+                    if !generic_trait_constraints_trait_names
+                        .contains(&structure_trait_constraint.trait_name)
+                    {
+                        errors.push(CompileError::TraitConstraintMissing {
+                            param: structure_type_info_with_engines.to_string(),
+                            trait_name: structure_trait_constraint.trait_name.suffix.to_string(),
+                            span: span.clone(),
+                        });
+                    }
+                }
+            } else {
+                // TODO(Esdrubal): Use callpath directly instead of suffix,
```

Comment:
Does  this mean that we would miscompile on similarly named traits at different paths or is it just a performance improvement?

Reply:
Yes as it was similarly named traits could be used interchangeably to avoid the error. The most recent version uses the trait name with full path so this should not occur.